File: Client/reasn-client/packages/common/models/EventDto.ts
import ModelMappingError from "@reasn/common/errors/ModelMappingError";
import { TagDtoSchema } from "@reasn/common/models/TagDto";
import { EventStatus } from "@reasn/common/enums/modelsEnums";
import { z } from "zod";

export const EventDtoSchema = z.object({
  Name: z.string(),
  AddressId: z.number(),
  Description: z.string(),
  OrganizerId: z.number(),
  StartAt: z
    .string()
    .datetime({ offset: true })
    .or(z.date())
    .transform((arg) => new Date(arg)),
  EndAt: z
    .string()
    .datetime({ offset: true })
    .or(z.date())
    .transform((arg) => new Date(arg)),
  CreatedAt: z
    .string()
    .datetime({ offset: true })
    .or(z.date())
    .transform((arg) => new Date(arg)),
  UpdatedAt: z
    .string()
    .datetime({ offset: true })
    .or(z.date())
    .transform((arg) => new Date(arg)),
  Slug: z.string().nullable(),
  Status: z.nativeEnum(EventStatus),
  Tags: z.array(TagDtoSchema),
});

export type EventDto = z.infer<typeof EventDtoSchema>;

export const EventDtoMapper = {
  fromObject: (entity: object): EventDto => {
    const result = EventDtoSchema.safeParse(entity);
    if (!result.success) {
      throw new ModelMappingError(
        "EventDto",
        result.error.message,
        result.error,
      );
    }
    return result.data;
  },
  fromJSON: (jsonEntity: string): any => {
    if (!jsonEntity) {
      throw new ModelMappingError("EventDto", "Empty JSON string");
    }
    const result = EventDtoSchema.safeParse(JSON.parse(jsonEntity));
    if (!result.success) {
      throw new ModelMappingError(
        "EventDto",
        result.error.message,
        result.error,
      );
    }
    return result.data;
  },
};

File: Server/ReasnAPI/ReasnAPI/Models/DTOs/EventDto.cs
using ReasnAPI.Models.Enums;

namespace ReasnAPI.Models.DTOs
{
    public class EventDto
    {
        public string Name { get; set; } = null!;
        public int AddressId { get; set; }
        public string Description { get; set; } = null!;
        public int OrganizerId { get; set; }
        public DateTime StartAt { get; set; }
        public DateTime EndAt { get; set; }
        public DateTime CreatedAt { get; set; }
        public DateTime UpdatedAt { get; set; }
        public string Slug { get; set; } = null!;
        public EventStatus Status { get; set; }
        public List<TagDto>? Tags { get; set; }
        public List<ParameterDto>? Parameters { get; set; }
    }
}